File: android-app/app/src/main/java/com/lifeplus/healthcare/network/AuthInterceptor.java
package com.lifeplus.healthcare.network;

import com.lifeplus.healthcare.util.SessionManager;
import okhttp3.Interceptor;
import okhttp3.Request;
import okhttp3.Response;
import java.io.IOException;

/**
 * Attaches the JWT Bearer token to every outgoing request automatically.
 * If the server returns 401, attempts a token refresh once, then retries.
 */
public class AuthInterceptor implements Interceptor {

    private final SessionManager session;

    public AuthInterceptor(SessionManager session) {
        this.session = session;
    }

    @Override
    public Response intercept(Chain chain) throws IOException {
        Request original = chain.request();

        // Skip auth header for auth endpoints
        String url = original.url().toString();
        if (url.contains("/auth/login") || url.contains("/auth/register")
                || url.contains("/auth/refresh")) {
            return chain.proceed(original);
        }

        String token = session.getAccessToken();
        Request request = token != null
                ? original.newBuilder()
                    .header("Authorization", "Bearer " + token)
                    .build()
                : original;

        Response response = chain.proceed(request);

        // 401 -> try refresh once
        if (response.code() == 401) {
            synchronized (this) {
                if (session.getRefreshToken() != null) {
                    response.close();
                    String newToken = TokenRefresher.refresh(session);
                    if (newToken != null) {
                        Request retried = original.newBuilder()
                                .header("Authorization", "Bearer " + newToken)
                                .build();
                        return chain.proceed(retried);
                    } else {
                        // Refresh failed, clear session and force logout logic could go here
                        session.clearSession();
                    }
                }
            }
        }
        return response;
    }
}

File: android-app/app/src/main/java/com/lifeplus/healthcare/data/api/AuthInterceptor.kt
package com.lifeplus.healthcare.data.api

import com.lifeplus.healthcare.data.local.SessionDataStore
import kotlinx.coroutines.flow.firstOrNull
import kotlinx.coroutines.runBlocking
import okhttp3.Interceptor
import okhttp3.Response
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Injects the JWT Bearer token into every outgoing request.
 *
 * NOTE: runBlocking is intentional here — OkHttp interceptors run on a
 * background I/O thread (never the main thread), so blocking that thread
 * briefly to read DataStore is safe and avoids the complexity of a
 * suspending interceptor.  The DataStore read is typically < 1 ms once
 * the preferences file is cached in memory.
 */
@Singleton
class AuthInterceptor @Inject constructor(
    private val session: SessionDataStore
) : Interceptor {

    override fun intercept(chain: Interceptor.Chain): Response {
        val token = runBlocking { session.accessToken.firstOrNull() }

        val request = chain.request().newBuilder().apply {
            if (!token.isNullOrBlank()) {
                addHeader("Authorization", "Bearer $token")
            }
            // Always send JSON content type for non-GET requests
            if (chain.request().method != "GET") {
                addHeader("Accept", "application/json")
            }
        }.build()

        val response = chain.proceed(request)

        // If 401 and we had a token, the token is expired — clear it
        if (response.code == 401 && !token.isNullOrBlank()) {
            runBlocking { session.clearSession() }
        }

        return response
    }
}
